from PyQt5 import QtCore, QtGui, QtWidgets
from pyqtgraph import PlotWidget, BarGraphItem
from plotters import PlotAcceleration
from Database import Database
from Communication import Communication
from random import randint
import resources_rc, qtawesome as qta, pyqtgraph as pg, numpy as np

class Ui_MainWindow(object):
    def setupUi(self, MainWindow: QtWidgets.QMainWindow):
        # Configurations
        MainWindow.setWindowFlag(QtCore.Qt.FramelessWindowHint)
        QtGui.QFontDatabase.addApplicationFont("resources\\Anurati-Regular.otf")
        pg.setConfigOption('background', (33, 33, 33))
        pg.setConfigOption('foreground', (197, 198, 199))
        self.database = Database()
        self.serial_connection = Communication()
        
        # Ploatter
        self.acceleration = PlotAcceleration()
        
        timer = QtCore.QTimer(MainWindow)
        timer.timeout.connect(lambda: self.update())
        timer.start(200)
    
        MainWindow.setObjectName("MainWindow")
        MainWindow.resize(794, 600)
        MainWindow.setStyleSheet("")
        self.centralwidget = QtWidgets.QWidget(MainWindow)
        self.centralwidget.setObjectName("centralwidget")
        self.verticalLayout = QtWidgets.QVBoxLayout(self.centralwidget)
        self.verticalLayout.setObjectName("verticalLayout")
        self.title_horizontalLayout = QtWidgets.QHBoxLayout()
        self.title_horizontalLayout.setSizeConstraint(QtWidgets.QLayout.SetDefaultConstraint)
        self.title_horizontalLayout.setObjectName("title_horizontalLayout")
        self.icon_label = QtWidgets.QLabel(self.centralwidget)
        sizePolicy = QtWidgets.QSizePolicy(QtWidgets.QSizePolicy.MinimumExpanding, QtWidgets.QSizePolicy.Preferred)
        sizePolicy.setHorizontalStretch(0)
        sizePolicy.setVerticalStretch(0)
        sizePolicy.setHeightForWidth(self.icon_label.sizePolicy().hasHeightForWidth())
        self.icon_label.setSizePolicy(sizePolicy)
        self.icon_label.setMinimumSize(QtCore.QSize(40, 40))
        self.icon_label.setMaximumSize(QtCore.QSize(40, 40))
        self.icon_label.setCursor(QtGui.QCursor(QtCore.Qt.PointingHandCursor))
        self.icon_label.setText("")
        self.icon_label.setPixmap(QtGui.QPixmap(":/resources/icon.png"))
        self.icon_label.setScaledContents(True)
        self.icon_label.setObjectName("icon_label")
        self.title_horizontalLayout.addWidget(self.icon_label)
        self.title_label = QtWidgets.QLabel(self.centralwidget)
        font = QtGui.QFont()
        font.setFamily("Anurati")
        font.setPointSize(15)
        font.setBold(True)
        self.title_label.setFont(font)
        self.title_label.setCursor(QtGui.QCursor(QtCore.Qt.PointingHandCursor))
        self.title_label.setObjectName("title_label")
        self.title_horizontalLayout.addWidget(self.title_label)
        self.options_layout = QtWidgets.QHBoxLayout()
        self.options_layout.setObjectName("options_layout")
        self.changeStateBtn = QtWidgets.QPushButton(self.centralwidget)
        sizePolicy = QtWidgets.QSizePolicy(QtWidgets.QSizePolicy.Fixed, QtWidgets.QSizePolicy.Fixed)
        sizePolicy.setHorizontalStretch(0)
        sizePolicy.setVerticalStretch(0)
        sizePolicy.setHeightForWidth(self.changeStateBtn.sizePolicy().hasHeightForWidth())
        self.changeStateBtn.setSizePolicy(sizePolicy)
        self.changeStateBtn.setMaximumSize(QtCore.QSize(30, 30))
        self.changeStateBtn.setCursor(QtGui.QCursor(QtCore.Qt.PointingHandCursor))
        self.changeStateBtn.setObjectName("changeStateBtn")
        icon = qta.icon("msc.chrome-maximize")
        self.changeStateBtn.setIcon(icon)
        self.changeStateBtn.clicked.connect(self.changeState)
        self.options_layout.addWidget(self.changeStateBtn)
        self.closeWindowBtn = QtWidgets.QPushButton(self.centralwidget, )
        sizePolicy = QtWidgets.QSizePolicy(QtWidgets.QSizePolicy.Fixed, QtWidgets.QSizePolicy.Fixed)
        sizePolicy.setHorizontalStretch(0)
        sizePolicy.setVerticalStretch(0)
        sizePolicy.setHeightForWidth(self.closeWindowBtn.sizePolicy().hasHeightForWidth())
        self.closeWindowBtn.setSizePolicy(sizePolicy)
        self.closeWindowBtn.setMaximumSize(QtCore.QSize(30, 30))
        self.closeWindowBtn.setCursor(QtGui.QCursor(QtCore.Qt.PointingHandCursor))
        self.closeWindowBtn.setObjectName("closeWindowBtn")
        self.closeWindowBtn.clicked.connect(self.closeWindow)
        icon = qta.icon("msc.close")
        self.closeWindowBtn.setIcon(icon)
        self.options_layout.addWidget(self.closeWindowBtn)
        self.title_horizontalLayout.addLayout(self.options_layout)
        self.verticalLayout.addLayout(self.title_horizontalLayout)
        self.viewLayouts = QtWidgets.QGridLayout()
        self.viewLayouts.setObjectName("viewLayouts")
        self.mainFrame = QtWidgets.QFrame(self.centralwidget)
        sizePolicy = QtWidgets.QSizePolicy(QtWidgets.QSizePolicy.Preferred, QtWidgets.QSizePolicy.Preferred)
        sizePolicy.setHorizontalStretch(0)
        sizePolicy.setVerticalStretch(0)
        sizePolicy.setHeightForWidth(self.mainFrame.sizePolicy().hasHeightForWidth())
        self.mainFrame.setSizePolicy(sizePolicy)
        self.mainFrame.setMinimumSize(QtCore.QSize(400, 400))
        self.mainFrame.setMaximumSize(QtCore.QSize(16777215, 16777215))
        self.mainFrame.setFrameShape(QtWidgets.QFrame.StyledPanel)
        self.mainFrame.setFrameShadow(QtWidgets.QFrame.Raised)
        self.mainFrame.setObjectName("mainFrame")
        self.gridLayout_2 = QtWidgets.QGridLayout(self.mainFrame)
        self.gridLayout_2.setContentsMargins(0, 0, 0, 0)
        self.gridLayout_2.setObjectName("gridLayout_2")
        self.scrollArea = QtWidgets.QScrollArea(self.mainFrame)
        self.scrollArea.setWidgetResizable(True)
        self.scrollArea.setObjectName("scrollArea")
        self.scrollAreaWidgetContents = QtWidgets.QWidget()
        self.scrollAreaWidgetContents.setGeometry(QtCore.QRect(0, -353, 753, 930))
        self.scrollAreaWidgetContents.setObjectName("scrollAreaWidgetContents")
        self.gridLayout_3 = QtWidgets.QGridLayout(self.scrollAreaWidgetContents)
        self.gridLayout_3.setObjectName("gridLayout_3")
        
        self.graphicsView_3 = PlotWidget(self.scrollAreaWidgetContents, title="Accelerations (m/sÂ²)")
        sizePolicy = QtWidgets.QSizePolicy(QtWidgets.QSizePolicy.Fixed, QtWidgets.QSizePolicy.Fixed)
        sizePolicy.setHorizontalStretch(0)
        sizePolicy.setVerticalStretch(0)
        sizePolicy.setHeightForWidth(self.graphicsView_3.sizePolicy().hasHeightForWidth())
        self.graphicsView_3.setSizePolicy(sizePolicy)
        self.graphicsView_3.setMinimumSize(QtCore.QSize(300, 300))
        self.graphicsView_3.setMaximumSize(QtCore.QSize(500, 500))
        self.graphicsView_3.setObjectName("graphicsView_3")
        self.gridLayout_3.addWidget(self.graphicsView_3, 1, 0, 1, 1)
        
        self.graphicsView_3.addLegend()
        self.graphicsView_3.hideAxis("bottom")
        self.x_axis = self.graphicsView_3.plot(pen=(102, 252, 241), name='X')
        self.y_axis = self.graphicsView_3.plot(pen=(29, 185, 84), name='Y')
        self.z_axis = self.graphicsView_3.plot(pen=(203, 45, 111), name='Z')
        self.graphicsView_3.setLabel(axis='left', text="Height")
        self.acc_x_data = np.linspace(0, 0)
        self.acc_y_data = np.linspace(0, 0)
        self.acc_z_data = np.linspace(0, 0)
        self.ptr = 0
        
        self.graphicsView_4 = PlotWidget(self.scrollAreaWidgetContents)
        sizePolicy = QtWidgets.QSizePolicy(QtWidgets.QSizePolicy.Fixed, QtWidgets.QSizePolicy.Fixed)
        sizePolicy.setHorizontalStretch(0)
        sizePolicy.setVerticalStretch(0)
        sizePolicy.setHeightForWidth(self.graphicsView_4.sizePolicy().hasHeightForWidth())
        self.graphicsView_4.setSizePolicy(sizePolicy)
        self.graphicsView_4.setMinimumSize(QtCore.QSize(300, 300))
        self.graphicsView_4.setMaximumSize(QtCore.QSize(500, 500))
        self.graphicsView_4.setObjectName("graphicsView_4")
        self.bargraphItem = BarGraphItem(x = [0], height = [0], width = 0.6, brush = 'r')
        self.graphicsView_4.addItem(self.bargraphItem)
        
        
        self.gridLayout_3.addWidget(self.graphicsView_4, 1, 1, 1, 1)
        self.graphicsView = PlotWidget(self.scrollAreaWidgetContents)
        sizePolicy = QtWidgets.QSizePolicy(QtWidgets.QSizePolicy.Fixed, QtWidgets.QSizePolicy.Fixed)
        sizePolicy.setHorizontalStretch(0)
        sizePolicy.setVerticalStretch(0)
        sizePolicy.setHeightForWidth(self.graphicsView.sizePolicy().hasHeightForWidth())
        self.graphicsView.setSizePolicy(sizePolicy)
        self.graphicsView.setMinimumSize(QtCore.QSize(300, 300))
        self.graphicsView.setMaximumSize(QtCore.QSize(500, 500))
        self.graphicsView.setObjectName("graphicsView")
        self.gridLayout_3.addWidget(self.graphicsView, 0, 1, 1, 1)
        self.graphicsView_2 = PlotWidget(self.scrollAreaWidgetContents)
        sizePolicy = QtWidgets.QSizePolicy(QtWidgets.QSizePolicy.Fixed, QtWidgets.QSizePolicy.Fixed)
        sizePolicy.setHorizontalStretch(0)
        sizePolicy.setVerticalStretch(0)
        sizePolicy.setHeightForWidth(self.graphicsView_2.sizePolicy().hasHeightForWidth())
        self.graphicsView_2.setSizePolicy(sizePolicy)
        self.graphicsView_2.setMinimumSize(QtCore.QSize(300, 300))
        self.graphicsView_2.setMaximumSize(QtCore.QSize(500, 500))
        self.graphicsView_2.setObjectName("graphicsView_2")
        self.gridLayout_3.addWidget(self.graphicsView_2, 0, 0, 1, 1)
        self.graphicsView_5 = PlotWidget(self.scrollAreaWidgetContents)
        sizePolicy = QtWidgets.QSizePolicy(QtWidgets.QSizePolicy.Fixed, QtWidgets.QSizePolicy.Fixed)
        sizePolicy.setHorizontalStretch(0)
        sizePolicy.setVerticalStretch(0)
        sizePolicy.setHeightForWidth(self.graphicsView_5.sizePolicy().hasHeightForWidth())
        self.graphicsView_5.setSizePolicy(sizePolicy)
        self.graphicsView_5.setMinimumSize(QtCore.QSize(300, 300))
        self.graphicsView_5.setMaximumSize(QtCore.QSize(500, 500))
        self.graphicsView_5.setObjectName("graphicsView_5")
        self.gridLayout_3.addWidget(self.graphicsView_5, 2, 0, 1, 1)
        self.graphicsView_6 = PlotWidget(self.scrollAreaWidgetContents)
        sizePolicy = QtWidgets.QSizePolicy(QtWidgets.QSizePolicy.Fixed, QtWidgets.QSizePolicy.Fixed)
        sizePolicy.setHorizontalStretch(0)
        sizePolicy.setVerticalStretch(0)
        sizePolicy.setHeightForWidth(self.graphicsView_6.sizePolicy().hasHeightForWidth())
        self.graphicsView_6.setSizePolicy(sizePolicy)
        self.graphicsView_6.setMinimumSize(QtCore.QSize(300, 300))
        self.graphicsView_6.setMaximumSize(QtCore.QSize(500, 500))
        self.graphicsView_6.setObjectName("graphicsView_6")
        self.gridLayout_3.addWidget(self.graphicsView_6, 2, 1, 1, 1)
        self.scrollArea.setWidget(self.scrollAreaWidgetContents)
        self.gridLayout_2.addWidget(self.scrollArea, 2, 1, 1, 1)
        self.viewLayouts.addWidget(self.mainFrame, 0, 0, 1, 1)
        self.verticalLayout.addLayout(self.viewLayouts)
        MainWindow.setCentralWidget(self.centralwidget)
        
        self.retranslateUi(MainWindow)
        QtCore.QMetaObject.connectSlotsByName(MainWindow)
    def acceleration_update(self, ax, ay, az):
        self.acc_x_data[:-1] = self.acc_x_data[1:]
        self.acc_y_data[:-1] = self.acc_y_data[1:]
        self.acc_y_data[:-1] = self.acc_y_data[1:]
        
        # Replacing last-indexed dublicate value with new value
        self.acc_x_data[-1] = float(ax)
        self.acc_y_data[-1] = float(ay)
        self.acc_z_data[-1] = float(az)
        self.ptr += 1
        
        self.x_axis.setData(self.acc_x_data)
        self.y_axis.setData(self.acc_y_data)
        self.z_axis.setData(self.acc_z_data)
        
        self.x_axis.setPos(self.ptr, 0) # Increasing X-Axis of GraphView
        self.y_axis.setPos(self.ptr, 0)
        self.z_axis.setPos(self.ptr, 0)
    def changeState(self):
        if MainWindow.isMaximized():
            MainWindow.showNormal()
            icon = qta.icon("msc.chrome-maximize")
            self.changeStateBtn.setIcon(icon)
            return
        icon = qta.icon("msc.chrome-minimize")
        self.changeStateBtn.setIcon(icon)
        MainWindow.showMaximized()
        return
    def update(self):
        try:
            value_chain: list = self.serial_connection.getData()
            if len(value_chain) > 1:
                self.acceleration_update(*value_chain[:3])
                self.database.guardar(value_chain)
        except IndexError as e:
            print(str(e))
    def closeWindow(self):
        MainWindow.close()
    def retranslateUi(self, MainWindow: QtWidgets.QMainWindow):
        self._translate = QtCore.QCoreApplication.translate
        MainWindow.setWindowTitle(self._translate("MainWindow", "MainWindow"))
        self.title_label.setText(self._translate("MainWindow", "Graphical Representation of Sensor Data"))
        # self.changeStateBtn.setText(self._translate("MainWindow", icon))
        # self.closeWindowBtn.setText(self._translate("MainWindow", "PushButton"))

if __name__ == "__main__":
    import sys
    app = QtWidgets.QApplication(sys.argv)
    MainWindow = QtWidgets.QMainWindow()
    ui = Ui_MainWindow()
    ui.setupUi(MainWindow)
    MainWindow.show()
    sys.exit(app.exec_())
